nixops/resources/sns_topic.py

```diff
@@ -0,0 +1,181 @@
+    """State of an SNS topic."""
+
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    topic_name = nixops.util.attr_property("ec2.topicName", None)
+    access_key_id = nixops.util.attr_property("ec2.accessKeyId", None)
+    display_name = nixops.util.attr_property("ec2.topicDisplayName", None)
+    region = nixops.util.attr_property("ec2.region", None)
+    policy = nixops.util.attr_property("ec2.topicPolicy", None)
+    arn = nixops.util.attr_property("ec2.topicARN", None)
+    subscriptions = nixops.util.attr_property("ec2.topicSubscriptions", [],'json')
+
+    @classmethod
+    def get_type(cls):
+        return "sns-topic"
+
+
+    def __init__(self, depl, name, id):
+        nixops.resources.ResourceState.__init__(self, depl, name, id)
+        self._conn = None
+
+
+    def show_type(self):
+        s = super(SNSTopicState, self).show_type()
+        if self.region: s = "{0} [{1}]".format(s, self.region)
+        return s
+
+
+    @property
+    def resource_id(self):
+        return self.topic_name
+
+    def prefix_definition(self, attr):
+        return {('resources', 'snsTopics'): attr}
+
+    def get_physical_spec(self):
+        return {'arn': self.arn}
+
+    def get_definition_prefix(self):
+        return "resources.snsTopics."
+
+
+    def connect(self):
+        if self._conn: return
+        assert self.region
+        (access_key_id, secret_access_key) = nixops.ec2_utils.fetch_aws_secret_key(self.access_key_id)
+        self._conn = boto.sns.connect_to_region(
+            region_name=self.region, aws_access_key_id=access_key_id, aws_secret_access_key=secret_access_key)
+
+
+    def _destroy(self):
+        if self.state != self.UP: return
+        self.connect()
+        self.log("destroying SNS topic ‘{0}’...".format(self.topic_name))
+        self._conn.delete_topic(self.arn)
+        with self.depl._db:
+            self.state = self.MISSING
+            self.topic_name = None
+            self.region = None
+            self.policy = None
+            self.arn = None
+
+    def topic_exists(self,arn):
+        return True
+
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        self.access_key_id = defn.access_key_id or nixops.ec2_utils.get_access_key_id()
+        if not self.access_key_id:
+            raise Exception("please set ‘accessKeyId’, $EC2_ACCESS_KEY or $AWS_ACCESS_KEY_ID")
+
+        arn = self.arn
+        if self.state == self.UP and (self.topic_name != defn.topic_name or self.region != defn.region):
+            self.log("topic definition changed, recreating...")
+            self._destroy()
+            self._conn = None
+
+        self.region = defn.region
+        self.connect()
+
+        if self.arn == None:
+            self.log("creating SNS topic ‘{0}’...".format(defn.topic_name))
+            topic = self._conn.create_topic(defn.topic_name)
+            arn = topic.get('CreateTopicResponse').get('CreateTopicResult')['TopicArn']
```

Comment:
What happens if we're interrupted after the `create_topic` AWS call, but before `self.depl._db` is updated? Will that leak a topic? If so, is there a way to make it idempotent (e.g. using a creation token like for EC2 instance creation)?